plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/declarations/FirKotlinUClass.kt

```diff
@@ -111,8 +123,50 @@ class FirKotlinUClass(
     }
 
     override fun getMethods(): Array<UMethod> {
-        // TODO: Not yet implemented
-        return super.getMethods()
+        val hasPrimaryConstructor = ktClass?.hasPrimaryConstructor() ?: false
+        var secondaryConstructorsCount = 0
+
+        fun createUMethod(psiMethod: PsiMethod): UMethod {
+            return if (psiMethod is KtLightMethod && psiMethod.isConstructor) {
+                if (!hasPrimaryConstructor && secondaryConstructorsCount++ == 0)
+                    FirKotlinSecondaryConstructorWithInitializersUMethod(ktClass, psiMethod, this)
+                else
+                    FirKotlinConstructorUMethod(ktClass, psiMethod, this)
+            } else {
+                languagePlugin?.convertOpt(psiMethod, this) ?: reportConvertFailure(psiMethod)
+            }
+        }
+
+        fun isDelegatedMethod(psiMethod: PsiMethod) = psiMethod is KtLightMethod && psiMethod.isDelegated
+
+        val result = ArrayList<UMethod>(javaPsi.methods.size)
+        val handledKtDeclarations = mutableSetOf<PsiElement>()
+
+        for (lightMethod in javaPsi.methods) {
+            if (isDelegatedMethod(lightMethod)) continue
+            val uMethod = createUMethod(lightMethod)
+            result.add(uMethod)
+
+            // Ensure we pick the main Kotlin origin, not the auxiliary one
+            val kotlinOrigin = (lightMethod as? KtLightMethod)?.kotlinOrigin ?: uMethod.sourcePsi
+            handledKtDeclarations.addIfNotNull(kotlinOrigin)
+        }
+
+        val ktDeclarations: List<KtDeclaration> = run ktDeclarations@{
+            ktClass?.let { return@ktDeclarations it.declarations }
+            (javaPsi as? KtLightClassForFacade)?.let { facade ->
+                return@ktDeclarations facade.files.flatMap { file -> file.declarations }
+            }
+            emptyList()
+        }
+
+        ktDeclarations.asSequence()
+            .filterNot { handledKtDeclarations.contains(it) }
+            .mapNotNullTo(result) {
+                baseResolveProviderService.baseKotlinConverter.convertDeclaration(it, this, arrayOf(UElement::class.java)) as? UMethod
+            }
+
+        return result.toTypedArray()
```

Comment:
Can this part of code be commonized too? There seems nothing to be nothing fir-specific here

Reply:
There is nothing specific to FIR here, except for creation of constructor methods: use of `FirKotlinConstructorUMethod ` and `FirKotlinSecondaryConstructorWithInitializersUMethod `. There are something else that bother commonizing those abstractions, which will be explained in the other comment.

So, by abstracting the creation of constructors, we can commonize these.